### PR TITLE
Debugging fundamentals task 1 solution

### PR DESCRIPTION
**Pull request by beratmehmet**
- Sort and IndexOf methods fixed.
- Equals() and GetHashCode() methods overridden for Product Class

### CODE REVIEW

File: Debugging Fundamentals/Task1/Utilities.cs [hunk 1]
```diff
@@ -35,12 +38,16 @@ public static void Sort(int[] numbers)
         /// otherwise -1.</returns>
         public static int IndexOf(Product[] products, Predicate<Product> predicate)
         {
-            for (int i = 0; i < products.Length - 1; i++)
+            //_ = products ??  throw new ArgumentNullException(nameof(products));
```
Comment (RamanPanfilenka) [inline Debugging Fundamentals/Task1/Utilities.cs:41]: You can delete comments, now they unnecessary

File: Debugging Fundamentals/Task1/Product.cs [hunk 2]
```diff
@@ -11,5 +11,22 @@ public Product(string name, double price)
         public string Name { get; set; }
 
         public double Price { get; set; }
+
+        public override bool Equals(object obj)
```
Comment (RamanPanfilenka) [inline Debugging Fundamentals/Task1/Product.cs:15]: Could you please review and update your logic of Equals override?
https://learn.microsoft.com/en-us/previous-versions/ms173147(v=vs.90)?redirectedfrom=MSDN

File: Debugging Fundamentals/Task1/Product.cs [hunk 3]
```diff
@@ -11,5 +11,22 @@ public Product(string name, double price)
...
+        {
+            if (obj == null || GetType() != obj.GetType())
+            {
+                return false;
+            }
+
+            Product other = (Product)obj;
+            return Name == other.Name && Price == other.Price;
+
```
Comment (RamanPanfilenka) [inline Debugging Fundamentals/Task1/Product.cs:24]: I think this line can be removed

File: Debugging Fundamentals/Task1/Utilities.cs [hunk 4]
```diff
@@ -11,15 +11,18 @@ public static class Utilities
         public static void Sort(int[] numbers)
         {
             int temp;
+            //_ = numbers ?? throw new ArgumentNullException(nameof(numbers));
```
Comment (RamanPanfilenka) [inline Debugging Fundamentals/Task1/Utilities.cs:14]: Also can be removed for the now

**Review by RamanPanfilenka — CHANGES_REQUESTED** [review]

File: Debugging Fundamentals/Task1/Product.cs (hunk 3 above)
Comment (beratmehmet) [reply Debugging Fundamentals/Task1/Product.cs:42]: This line changed when I followed the guideline.

File: Debugging Fundamentals/Task1/Product.cs (hunk 2 above)
Comment (beratmehmet) [reply Debugging Fundamentals/Task1/Product.cs]: I followed the guideline and implemented two Equals method, one for System.Object type and one for Product type as recommended.

Comment (RamanPanfilenka) [reply Debugging Fundamentals/Task1/Product.cs]: Sorry I have send you outdated link, you need follow this one: https://learn.microsoft.com/en-us/visualstudio/ide/reference/generate-equals-gethashcode-methods?view=vs-2022
In your current implementation you have dublicated loginc for two Equals. But one of them should reuse logic from the second one

Comment (beratmehmet) [reply Debugging Fundamentals/Task1/Product.cs:17]: Oh, I see. I followed the new document and liked this approach more 😄

**Pull request merged**